README.txt
Requirements
************

Python > 2.5.2
http://www.python.org/download/releases

To use smtp authentication over ssl you have to configure python with SSL

* Required package libssl-dev (e.g Ubuntu dapper)
$ ./configure --with-ssl
$ make
$ make install


Test enviroment
***************
Ubuntu 6.06 Python 2.6 with SSL (self compiled)
Ubuntu 7.04 Python 2.5.2
Ubuntu 8.10 Python 2.6.5


Before you start
****************
* check the following vars in the linkcrawler.py class and change
them for your needs

* CONFIG_PATH = "config"
* CONFIG_FILE_NAME = "application.cfg"
* LOG_FILE = "/tmp/linkcrawler.out"

* check the default settings in the config file

* Execute test_linkcrawler.py to run a unittest


Execute App
***********
* by shell
$ python linkcrawler.py [SITENAME]

* by crontab
$ crontab -e // edit crontab
// Example execute app once a month
0      0       1       *       *       /path-to-the-app/linkcrawler.py [SITENAME] >/dev/null 2>&1


Know open issues
****************
* Add missing unit-tests
* set log level at command line LOG_LEVEL
* set log level for every site separatly
* checkExternalLink: doesn't save double external links
* processedExtern: should be a dictionary to save the last valid response code

changelog.txt
23-01-2011  Clemens Tietze  <clemens.tietze@iltisworld.de>
---
added: threading
added: MAX_THREADS
added: sitename-sample config

10-05-2010  Clemens Tietze  <clemens.tietze@iltisworld.de>
---
changed: completeRelativePath method now extracts the parent path

07-05-2010  Clemens Tietze  <clemens.tietze@iltisworld.de>
---
changed: Unittest method "test_path_regex" to "test_completeRelativePath"
Improved: new method completeRelativePath to make a server absolute path within the parent and the actual path

06-05-2010  Clemens Tietze  <clemens.tietze@iltisworld.de>
---
* Added Unittest method "test_extractLinks"
* Added Unittest method "test_path_regex"
* Improved ignoreLink method
* Improved extractLinks method to find relative links too
* Improved readDataRecursivly method to add the former path if the link is relative
* Improved loadConfigurationSite load INCLUDE_PATTERN / EXCLUDE_PATTERN checks now for empty string

25-03-2010  Clemens Tietze  <clemens.tietze@iltisworld.de>
---
* Added destructor
* Improved info and debug messages
* Added LINK_DEPTH option to loadConfiguration method

24-03-2010  Clemens Tietze  <clemens.tietze@iltisworld.de>
---
* Improved "smtpsend" to send mails without authentification
* Added Unittest method "test_write2file"
* Improved "ignoreLink" method and debug message
* Renamed option EXCLUDE_TYPES to EXCLUDE_PATTERN
* Renamed option INCLUDE_TYPES to INCLUDE_FILETYPES
* Renamed "checkLink" method to "ignoreLink"
* Renamed all config options
* Refactored "loadConfiguration" and splitted to "loadConfiguration" and "loadConfigurationSite"
* Added single config file for specific site
* Restructured the project file system

23-03-2010  Clemens Tietze  <clemens.tietze@iltisworld.de>
---
* Added close and connect httpconnection method to "start"
* Removed close and connect httpconnection from getResponse
* Added configuration option: useragent
* Added cookie to getResponse httpconnection
* Improved getResponse connection is now keep-alive
* Added configuration option: SMTP_TTLS
* Added smtpsend method
* Removed mailHelper class
* Added unittests
* Added configuration option: mail server settings
* Improved debug, exception and error messages

22-03-2010  Clemens Tietze  <clemens.tietze@iltisworld.de>
---
* Added configuration option: reportlevel
* Added showStatusCode
* Improvement: generateReport compares now the statuscode with the reportlevel setting

25-01-2010  Clemens Tietze  <clemens.tietze@iltisworld.de>
---
* Added sections to config file
* Added config file

src/test_linkcrawler.py
#!/usr/bin/python
'''
' $RCSfile: test_LinkCrawler.py,v $
' $Source: /home/cvs/sc_webreport/src/test_LinkCrawler.py,v $
' $Date: 2010-05-10 14:48:13 $
' $Revision: 1.4 $
' $Author: ctietze $
' Remarks:
'''

import unittest
import sys
from linkcrawler import LinkCrawler

from library.Iprods.Http.Utils import Iprods_Http_Utils
from library.Iprods.File.Utils import Iprods_File_Utils
from library.Iprods.Date.Utils import Iprods_Date_Utils
from library.Iprods.String.Utils import Iprods_String_Utils

from checker import Checker
from reader import Reader

class SimplisticTest(unittest.TestCase):
    #def __init__(self):

    def test_getInHMS(self):
        print ">>> test_getInHMS"
        app = LinkCrawler()
        app.LOG_LEVEL = 10
        timeInHMS = Iprods_Date_Utils.getInHMS(3600)

        print "<<< test_getInHMS [seconds: 3600, timeInHMS: %s]\n" % (timeInHMS)
        self.failUnless(timeInHMS)

    def test_getUrlStatus(self):
        print ">>> test_getUrlStatus"
        app     = LinkCrawler()
        checker = Checker(app)
        configIsLoaded = app.loadConfigurationSite("unittest")

        if configIsLoaded is False:
            self.fail("load configuration failed")
        else:
            statusCode = checker.getStatusCode("http://www.google.de")

        print "<<< test_getUrlStatus [statusCode: %s]\n" % statusCode
        self.failUnless(statusCode)

    def test_loadConfiguration(self):
        print ">>> test_loadConfiguration"
        app = LinkCrawler()
        configIsLoaded = app.loadConfigurationSite("unittest")

        print "<<< test_loadConfiguration [configIsLoaded: %s]\n" % configIsLoaded
        self.failUnless(configIsLoaded)

    def test_isCodeInCodeRange_fail(self):
        print ">>> test_isCodeInCodeRange_fail"

        codes = "100,200,500"
        showCode = Iprods_Http_Utils.isCodeInCodeRange(codes, "405")

        print "<<< test_isCodeInCodeRange_fail [range: %s, code: 405, showCode: %s]\n" % (codes, showCode)
        self.failIf(showCode)

    def test_isCodeInCodeRange_success(self):
        print ">>> test_isCodeInCodeRange_success"

        codes = "100,200,500"
        showCode = Iprods_Http_Utils.isCodeInCodeRange(codes, "405")

        print "<<< test_isCodeInCodeRange_success [range: %s, code: 405, showCode: %s]\n" % (codes, showCode)
        self.failUnless(showCode)

    def test_smtpsend(self):
        print ">>> test_smtpsend"
        app = LinkCrawler()
        configIsLoaded = app.loadConfigurationSite("unittest")

        if configIsLoaded is False:
            self.fail("load configuration failed")
        else:
            sendResult = app.smtpsend("clemens.tietze@gms.de", "clemens.tietze@gmx.de", "", "", "Webreport Unittest test_smtpsend", "Lorem ipsum")

        print "<<< test_smtpsend [sendResult: %s]\n" % sendResult
        self.failIfEqual(sendResult, None)

    def test_write2file(self):
        print ">>> test_write2file"
        writeResult = True
        file = ""

        try:
            app = LinkCrawler()
            file = "%s%s.txt" % (app.REPORT_PATH, Iprods_File_Utils.getValidFileName("unittest"))
            Iprods_File_Utils.write2file(file, "Webreport Unittest test_write2file")
        except:
            writeResult = False
            exctype, value = sys.exc_info()[:2]
            print("exctype: %s, value: %s"  %  (exctype, value))

        print "<<< test_write2file [file: %s]\n" % (file)
        self.failUnless(writeResult)

    def test_extractLinks(self):
        print ">>> test_extractLinks"
        app    = LinkCrawler()
        reader = Reader(app)
        configIsLoaded = app.loadConfigurationSite("unittest")

        if configIsLoaded is False:
            self.fail("load configuration failed")
        else:
            response     = reader.getResponse('http://www.scandio.de')
            responseData = response[3]
            links = reader.extractLinks(responseData, 'http://www.scandio.de')

        print "<<< test_extractLinks [links: %s]\n" % links
        self.failUnless(links)

    def test_completeRelativePath(self):
        print ">>> test_completeRelativePath\n"
        app        = LinkCrawler()
        reader     = Reader(app)

        array = [['/en/de/index.html', '/aa/bb/index.htm']
                , ['/en/de/index.html', 'cc/dd/index.htm']
                , ['/en/de/index.html', '../cc/dd/index.htm']
                , ['/en/', 'search_iframe_en.htm']
                , ['/en/', '/search_iframe_en.htm']
                , ['/en/', '../search_iframe_en.htm']]

        for i in range(0, len(array)):
            relPath    = reader.completeRelativePath(array[i][0], array[i][1])
            print "test_completeRelativePath [path: %s, parent: %s, relPath: %s]\n" % (array[i][0],array[i][1],relPath)
            self.failUnless(relPath)

        print "<<< test_completeRelativePath \n"


if __name__ == '__main__':
    unittest.main()

src/checker.py
###
# $Date: 2010-05-06 16:26:06 $
# $Revision: 1.2 $
# $Author: ctietze $
# Remarks:
# checks all external links which are in the queue
###

import threading
import Queue
import httplib
import sys
import urlparse
import time

class Checker(threading.Thread):
    result    = {}
    lock      = threading.Lock()
    queue     = Queue.Queue()
    reportApp = None

    def __init__(self, reportApp):
        threading.Thread.__init__(self)

        self.reportApp = reportApp
        self.log = reportApp.getLogger()

    def run(self,):
        while True:

            if self.reportApp.TIMEOUT:
                time.sleep(self.reportApp.TIMEOUT)

            url  = Checker.queue.get()
            code = self.getStatusCode(url)
            Checker.lock.acquire()
            Checker.result[url] = code
            Checker.lock.release()
            Checker.queue.task_done()

    '''
    get status code
    '''
    def getStatusCode(self, url):
        self.log.debug(">>> getStatusCode [url: %s]" % (url))
        url = url.strip()
        (scheme, netloc, path, params, query, fragment) = urlparse.urlparse(url)
        status = 400

        if netloc != '':
            try:
                if sys.version_info < (2,6):
                    conn = httplib.HTTPConnection(netloc, 80)
                else:
                    conn = httplib.HTTPConnection(netloc, 80, timeout=10)

                conn.request("HEAD", path)
                res    = conn.getresponse()
                status = res.status
                conn.close()
            except:
                exctype, value = sys.exc_info()[:2]
                self.log.exception("getStatusCode failed. [netloc: %s, path: %s, exctype: %s, value: %s]  "  %  (netloc, path, exctype, value))

        self.log.debug("<<< getStatusCode [status: %s]" % (status))
        return status

src/linkcrawler.py
#!/usr/bin/python
'''
' $Date: 2010-05-10 14:48:13 $
' $Revision: 1.6 $
' $Author: ctietze $
' Remarks:
'''

import ConfigParser
import cStringIO
from email import encoders
import logging
import os
import smtplib, email
import sys

import time
from time import strftime

from library.Iprods.Http.Utils import Iprods_Http_Utils
from library.Iprods.File.Utils import Iprods_File_Utils
from library.Iprods.Date.Utils import Iprods_Date_Utils
from library.Iprods.String.Utils import Iprods_String_Utils

from checker import Checker
from reader import Reader

class LinkCrawler:

    CONFIG_PATH = "config"
    CONFIG_FILE_NAME = "application.cfg"

    LOG_FILE = "/tmp/linkcrawler.out"
    LOG_FORMAT = "%(asctime)s %(levelname)s: %(message)s"
    LOG_DATE = "%d.%m.%Y %H:%M:%S"
    LOG_LEVEL = logging.INFO

    countExcluded   = 0
    countDouble     = 0
    elapsedTime     = 0
    lastReader      = -1

    cookie = None
    connection = None

    processedIntern = []
    failedIntern    = []
    checkedIntern   = {}
    processedExtern = []
    statusCodes     = {}
    result          = {}

    log = logging.getLogger("LinkCrawler")

    '''
    ${classdocs}
    '''
    def __init__(self, sitename=None):
        logging.basicConfig(filename=self.LOG_FILE,level=self.LOG_LEVEL,format =self.LOG_FORMAT ,datefmt = self.LOG_DATE)
        self.log.debug(">> __init__")
        self.log.info("Start crawling....")

        configIsLoaded = self.loadConfiguration()

        if configIsLoaded:
            if sitename is not None:
                siteIsLoaded = self.loadConfigurationSite(sitename)
                if siteIsLoaded:
                    self.run()
            else:
                self.log.info("No sitename passed... waiting for input...")
        self.log.debug("<< __init__")

    '''
    load application configuration
    '''
    def loadConfiguration(self):
        self.log.debug(">>> loadConfiguration")

        try:
            config = ConfigParser.ConfigParser()
            configfile = os.getcwd( ) + "/" + self.CONFIG_PATH + "/" + self.CONFIG_FILE_NAME
            config.readfp(open(configfile))

            # read default configs
            self.SEND_REPORT    = Iprods_String_Utils.smart_bool(config.get('application', 'SEND_REPORT'))
            self.WRITE_REPORT   = Iprods_String_Utils.smart_bool(config.get('application', 'WRITE_REPORT'))
            self.VERBOSE_REPORT = Iprods_String_Utils.smart_bool(config.get('application', 'VERBOSE_REPORT'))
            self.LINK_DEPTH     = config.getint('application', 'LINK_DEPTH')
            self.REPORT_PATH    = config.get('application', 'REPORT_PATH')
            self.REPORT_LEVEL   = config.get('application', 'REPORT_LEVEL')
            self.USERAGENT      = config.get('application', 'USERAGENT')
            self.CHECK_EXTERNAL = config.get('application', 'CHECK_EXTERNAL')
            self.MAX_THREADS = config.getint('application', 'MAX_THREADS')

            self.TIMEOUT      = config.getfloat('application', 'TIMEOUT')
            self.HTTP_TIMEOUT = config.getint('application', 'HTTP_TIMEOUT')

            self.SMTP_SERVER   = config.get('application', 'SMTP_SERVER')
            self.SMTP_TTLS     = config.get('application', 'SMTP_TTLS')
            self.SMTP_PORT     = config.get('application', 'SMTP_PORT')
            self.SMTP_USERNAME = config.get('application', 'SMTP_USERNAME')
            self.SMTP_PASSWORD = config.get('application', 'SMTP_PASSWORD')

            self.MAIL_SUBJECT = config.get('application', 'MAIL_SUBJECT')
            self.MAIL_TO      = config.get('application', 'MAIL_TO')
            self.MAIL_FROM    = config.get('application', 'MAIL_FROM')
            self.MAIL_CC      = config.get('application', 'MAIL_CC')
            self.MAIL_BCC     = config.get('application', 'MAIL_BCC')

            self.log.info("application config loaded successfully... [SEND_REPORT: %s, WRITE_REPORT: %s, VERBOSE_REPORT: %s, LINK_DEPTH: %s, REPORT_PATH: %s, REPORT_LEVEL: %s, USERAGENT: %s, TIMEOUT: %s, HTTP_TIMEOUT: %s, SMTP_SERVER: %s, SMTP_TTLS: %s, SMTP_PORT: %s, SMTP_USERNAME: %s, MAIL_SUBJECT: %s, MAIL_TO: %s, MAIL_FROM: %s, MAIL_BCC: %s]" % (self.SEND_REPORT, self.WRITE_REPORT, self.VERBOSE_REPORT, self.LINK_DEPTH, self.REPORT_PATH, self.REPORT_LEVEL, self.USERAGENT, self.TIMEOUT, self.HTTP_TIMEOUT, self.SMTP_SERVER, self.SMTP_TTLS, self.SMTP_PORT, self.SMTP_USERNAME, self.MAIL_SUBJECT, self.MAIL_TO, self.MAIL_FROM, self.MAIL_BCC))

        except StandardError:
            config = None
            exctype, value = sys.exc_info()[:2]
            self.log.exception("loadConfiguration failed. exctype: %s, value: %s [file: %s]"  %  (exctype, value, self.CONFIG_FILE_NAME))

        self.log.debug("<<< loadConfiguration [config: %s]", config)
        return config

    def loadConfigurationSite(self, sitename):
        config = ''
        self.log.debug(">>> loadConfigurationSite [sitename: %s]" % (sitename))

        if sitename:
            try:
                config = ConfigParser.ConfigParser()
                configfile = os.getcwd( ) + "/" + self.CONFIG_PATH + "/" + sitename + ".cfg"
                config.readfp(open(configfile))

                if (config.has_section(sitename)):
                    # read site configs
                    self.DOMAIN = config.get(sitename, 'DOMAIN')
                    self.PATH = config.get(sitename, 'PATH')

                    if (config.has_option(sitename, 'MAX_THREADS')):
                        self.MAX_THREADS = config.getint(sitename, 'MAX_THREADS')
                    if (config.has_option(sitename, 'LINK_DEPTH')):
                        self.LINK_DEPTH = config.getint(sitename, 'LINK_DEPTH')
                    if (config.has_option(sitename, 'REPORT_LEVEL')):
                        self.REPORT_LEVEL = config.get(sitename, 'REPORT_LEVEL')
                    if (config.has_option(sitename, 'EXCLUDE_PATTERN') and len(config.get(sitename, 'EXCLUDE_PATTERN').strip()) > 0):
                        self.EXCLUDE_PATTERN = config.get(sitename, 'EXCLUDE_PATTERN').strip().split(",")
                    else:
                        self.EXCLUDE_PATTERN = ''
                    if (config.has_option(sitename, 'INCLUDE_FILETYPES') and len(config.get(sitename, 'INCLUDE_FILETYPES').strip()) > 0):
                        self.INCLUDE_FILETYPES = config.get(sitename, 'INCLUDE_FILETYPES').strip().split(",")
                    else:
                        self.INCLUDE_FILETYPES = ''

                    if (config.has_option(sitename, 'CHECK_EXTERNAL')):
                        self.CHECK_EXTERNAL = Iprods_String_Utils.smart_bool(config.get(sitename, 'CHECK_EXTERNAL'))
                    if (config.has_option(sitename, 'SEND_REPORT')):
                        self.SEND_REPORT = Iprods_String_Utils.smart_bool(config.get(sitename, 'SEND_REPORT'))
                    if (config.has_option(sitename, 'WRITE_REPORT')):
                        self.WRITE_REPORT = Iprods_String_Utils.smart_bool(config.get(sitename, 'WRITE_REPORT'))
                    if (config.has_option(sitename, 'VERBOSE_REPORT')):
                        self.VERBOSE_REPORT = Iprods_String_Utils.smart_bool(config.get(sitename, 'VERBOSE_REPORT'))

                    if (config.has_option(sitename, 'TIMEOUT')):
                        self.TIMEOUT = config.getfloat(sitename, 'TIMEOUT')
                    if (config.has_option(sitename, 'HTTP_TIMEOUT')):
                        self.HTTP_TIMEOUT = config.getint(sitename, 'HTTP_TIMEOUT')

                    if (config.has_option(sitename, 'SMTP_SERVER')):
                        self.SMTP_SERVER = config.get(sitename, 'SMTP_SERVER')
                    if (config.has_option(sitename, 'SMTP_TTLS')):
                        self.SMTP_TTLS = config.get(sitename, 'SMTP_TTLS')
                    if (config.has_option(sitename, 'SMTP_PORT')):
                        self.SMTP_PORT = config.get(sitename, 'SMTP_PORT')
                    if (config.has_option(sitename, 'SMTP_USERNAME')):
                        self.SMTP_USERNAME = config.get(sitename, 'SMTP_USERNAME')
                    if (config.has_option(sitename, 'SMTP_PASSWORD')):
                        self.SMTP_PASSWORD = config.get(sitename, 'SMTP_PASSWORD')
                    if (config.has_option(sitename, 'MAIL_SUBJECT')):
                        self.MAIL_SUBJECT = "%s - %s" % (config.get(sitename, 'MAIL_SUBJECT'), self.DOMAIN)
                    if (config.has_option(sitename, 'MAIL_FROM')):
                        self.MAIL_FROM = config.get(sitename, 'MAIL_FROM')
                    if (config.has_option(sitename, 'MAIL_TO')):
                        self.MAIL_TO = config.get(sitename, 'MAIL_TO')
                    if (config.has_option(sitename, 'MAIL_CC')):
                        self.MAIL_CC = config.get(sitename, 'MAIL_CC')
                    if (config.has_option(sitename, 'MAIL_BCC')):
                        self.MAIL_BCC = config.get(sitename, 'MAIL_BCC')

                    self.log.info("site configuration loaded successfully... [sitename: %s, SEND_REPORT: %s, WRITE_REPORT: %s, VERBOSE_REPORT: %s, LINK_DEPTH: %s, REPORT_PATH: %s, REPORT_LEVEL: %s, USERAGENT: %s, TIMEOUT: %s, HTTP_TIMEOUT: %s, SMTP_SERVER: %s, SMTP_TTLS: %s, SMTP_PORT: %s, SMTP_USERNAME: %s, MAIL_SUBJECT: %s, MAIL_TO: %s, MAIL_FROM: %s, MAIL_BCC: %s]" % (sitename, self.SEND_REPORT, self.WRITE_REPORT, self.VERBOSE_REPORT, self.LINK_DEPTH, self.REPORT_PATH, self.REPORT_LEVEL, self.USERAGENT, self.TIMEOUT, self.HTTP_TIMEOUT, self.SMTP_SERVER, self.SMTP_TTLS, self.SMTP_PORT, self.SMTP_USERNAME, self.MAIL_SUBJECT, self.MAIL_TO, self.MAIL_FROM, self.MAIL_BCC))

                else:
                    config = None
                    self.log.error("loadConfigurationSite failed. [sitename:%s]. " % (sitename))
            except StandardError:
                config = None
                exctype, value = sys.exc_info()[:2]
                self.log.exception("loadConfigurationSite failed. exctype: %s, value: %s [sitename: %s]"  %  (exctype, value, sitename))

        self.log.debug("<<< loadConfigurationSite [config: %s]", config)
        return config


    def run(self):
        self.log.debug(">>> run [DOMAIN: %s, path: %s]" % (self.DOMAIN, self.PATH))

        try:
            start     = time.time()
            printTime = start + 30
            isFinished = False

            reader_threads = [Reader(self) for i in range(self.MAX_THREADS)]
            for thread in reader_threads:
                thread.setDaemon(True)
                thread.start()

            if self.CHECK_EXTERNAL is True:
                checker_threads = [Checker(self) for i in range(self.MAX_THREADS)]
                for thread in checker_threads:
                    thread.setDaemon(True)
                    thread.start()

            Reader.lock.acquire()
            Reader.result[self.PATH] = "busy"
            Reader.lock.release()
            Reader.queue.put(self.PATH)

            while True:
                actTime  = time.time()
                lastTime = actTime - self.lastReader
                #either queues are empty or reached max idle time


                if round(printTime,0) <=  round(actTime,0):
                    Reader.lock.acquire()
                    lenReader = len(Reader.result)
                    Reader.lock.release()

                    Checker.lock.acquire()
                    lenChecker = len(Checker.result)
                    Checker.lock.release()
                    print "status [links-read: %s, external-links: %s]" % (lenReader, lenChecker)
                    printTime = actTime + 15

                if (lastTime > 30)\
                    and Reader.queue.empty() and (self.CHECK_EXTERNAL is False or Checker.queue.empty()):
                #print "status [readerIsEmpty: %s, checkerIsEmpty: %s, lastReader: %s, lastTime: %s]" % (Reader.queue.empty(), Checker.queue.empty(), self.lastReader, lastTime)
                    break;

            Checker.queue.join()

            self.elapsedTime = (time.time() - start)

            if self.VERBOSE_REPORT or self.WRITE_REPORT or self.SEND_REPORT:
                report = self.generateReport('text')
                csv    = self.generateReport('csv')
                logFilename = Iprods_File_Utils.getValidFileName(self.DOMAIN, True)
                txtFile     = "%s%s.txt" % (self.REPORT_PATH, logFilename)
                csvFile     = "%s%s.csv" % (self.REPORT_PATH, logFilename)

                if self.VERBOSE_REPORT:
                    print report

                if self.WRITE_REPORT:
                    try:
                        Iprods_File_Utils.write2file(txtFile, report)
                    except:
                        exctype, value = sys.exc_info()[:2]
                        self.log.exception("write2file failed. exctype: %s, value: %s [filename: %s, report: %s]"  %  (exctype, value, txtFile, report))
                    try:
                        Iprods_File_Utils.write2file(csvFile, csv)
                    except:
                        exctype, value = sys.exc_info()[:2]
                        self.log.exception("write2file failed. exctype: %s, value: %s [filename: %s, csv: %s]"  %  (exctype, value, csvFile, csv))

                if self.SEND_REPORT:
                    self.smtpsend(self.MAIL_FROM, self.MAIL_CC, self.MAIL_BCC,\
                        self.MAIL_TO, self.MAIL_SUBJECT, report, csvFile)

        except StandardError:
            exctype, value = sys.exc_info()[:2]
            self.log.exception("run failed exc: %s, val: %s"  %  (exctype, value))

        self.log.debug("<<< run")

    '''
    generate the link report
    '''
    def generateReport(self, type):
        self.log.debug(">>> generateReport [type: %s]", type)
        data = cStringIO.StringIO()

        if type == 'text':
            data.write("Date: %s" % strftime("%Y-%m-%d %H:%M:%S"))
            data.write("\nDomain: %s" % self.DOMAIN)
            data.write("\nPath: %s" % self.PATH)
            data.write("\nShow Statuscodes: %s" % self.REPORT_LEVEL)
            if (self.EXCLUDE_PATTERN):
                data.write("\nExclude pattern: %s" % self.EXCLUDE_PATTERN)
            data.write("\nElapsed time: %s (h:m:s)" % Iprods_Date_Utils.getInHMS(self.elapsedTime))
            data.write("\nResult [internal: %s, external: %s, excluded: %s, double: %s]" % (len(self.processedIntern), len(self.processedExtern), self.countExcluded, self.countDouble))

            data.write("\n\nInternal Failed:")

            if len(self.failedIntern) > 0:
                for internal in self.failedIntern:
                    data.write("\n%s" % internal)

            data.write("\n\nExternal Failed:")

            if len(self.statusCodes) > 0:
                #{'200' : { 'URL' : [EXTERNALLINK1, EXTERNALLINK2 ....] }, '404' : {'URL' : [....]}
                #for key in sorted(self.statusCodes):
                codes = self.REPORT_LEVEL.split(',')
                for code in sorted(self.statusCodes, reverse=True):

                    Iprods_Http_Utils.isCodeInCodeRange(codes, code)
                    if self.showStatusCode(code):
                        data.write("\n\nHTTP_STATUS_CODE: %s\n" % code)
                        urls = self.statusCodes[code]
                        for url in urls:
                            data.write("\nURL: %s\n" % url)
                            for external in urls[url]:
                                data.write("EXT: %s\n" % external)


        if type == 'csv':
            if len(self.statusCodes) > 0:
                data.write("HTTP_STATUS_CODE, URL, EXTERN")
                #{'200' : { 'URL' : [EXTERNALLINK1, EXTERNALLINK2 ....] }, '404' : {'URL' : [....]}
                for code in sorted(self.statusCodes, reverse=True):
                    if self.showStatusCode(code):
                        urls = self.statusCodes[code]
                        for url in urls:
                            for external in urls[url]:
                                data.write("\n%s, %s, %s" % (code, url, external))

        #if len(self.processedExtern) > 0:
            #p.write("\n\nProcessed internal links\n %s" % str(self.processedIntern))
        #	data.write("\n\nProcessed external links\n%s" % str(self.processedExtern))

        report = data.getvalue()
        data.close()
        self.log.debug("<<< generateReport [report: %s]", report)
        return report

    def __del__(self):
        self.log.debug(">> __del__")

        # close connection if not already done
        #self.close()

        #shutdown logger
        self.log.info("Finished. Application stops now.")
        self.log.debug("<< __del__")
        logging.shutdown()

    def smtpsend(self, mailFrom, mailTo, mailCC, mailBCC, mailSubject, text, file=None):
        self.log.debug(">>> smtpsend [mailFrom: %s, mailTo: %s, mailCC: %s, mailBCC: %s, mailSubject: %s, text: %s]"  %  (mailFrom, mailTo, mailCC, mailBCC, mailSubject, text))
        result = None
        try:
            msg  = email.MIMEMultipart.MIMEMultipart()
            body = email.MIMEText.MIMEText(text)

            msg.attach(body)

            if file is not  None:
                attachment = email.MIMEBase.MIMEBase('text', 'plain')
                attachment.set_payload(open(file).read())
                attachment.add_header('Content-Disposition', 'attachment', filename=os.path.basename(file))
                encoders.encode_base64(attachment)
                msg.attach(attachment)

            msg.add_header('Subject', mailSubject)
            msg.add_header('From', mailFrom)
            msg.add_header('To', mailTo)
            if mailCC: msg['Cc'] = mailCC
            if mailBCC: msg['Bcc'] = mailBCC

            # now send the message
            server = smtplib.SMTP(self.SMTP_SERVER, self.SMTP_PORT)
            server.ehlo()

            if self.SMTP_TTLS:
                server.starttls()

            if self.SMTP_USERNAME and self.SMTP_PASSWORD:
                server.login(self.SMTP_USERNAME, self.SMTP_PASSWORD)

            result = server.sendmail(msg['From'], [mailTo, mailCC, mailBCC], msg.as_string())
            server.close()
        except :
            exctype, value = sys.exc_info()[:2]
            self.log.error("smtpsend failed. exctype: %s, value: %s [server: %s, port: %s, ttls: %s, user: %s] [mailFrom: %s, mailTo: %s, mailCC: %s, mailBCC: %s, mailSubject: %s, text: %s]"  %  (exctype, value, self.SMTP_SERVER, self.SMTP_PORT, self.SMTP_TTLS, self.SMTP_USERNAME, mailFrom, mailTo, mailCC, mailBCC, mailSubject, text))

        self.log.debug("<<< smtpsend [senderrs: %s]", result)
        return result
#
    def getLogger(self):
        return self.log

    def appendIntern(self, path):
        self.processedIntern.append(path)

if __name__ == '__main__':
    if len(sys.argv) > 1:
        LinkCrawler(sys.argv[1])
    else:
        LinkCrawler()
